Add Mean Absolute Percentage Error

pranjaldatta opened the pull request:
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #235.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

pep8speaks commented:
Hello @pranjaldatta! Thanks for opening this PR. 







* In the file [`torchmetrics/functional/regression/mean_absolute_percentage_error.py`](https://github.com/PyTorchLightning/metrics/blob/0f8138411360e65c9fbb4192a355205f40126cf1/torchmetrics/functional/regression/mean_absolute_percentage_error.py):

> [Line 25:121](https://github.com/PyTorchLightning/metrics/blob/0f8138411360e65c9fbb4192a355205f40126cf1/torchmetrics/functional/regression/mean_absolute_percentage_error.py#L25): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (127 > 120 characters)
> [Line 29:5](https://github.com/PyTorchLightning/metrics/blob/0f8138411360e65c9fbb4192a355205f40126cf1/torchmetrics/functional/regression/mean_absolute_percentage_error.py#L29): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 46:89](https://github.com/PyTorchLightning/metrics/blob/0f8138411360e65c9fbb4192a355205f40126cf1/torchmetrics/functional/regression/mean_absolute_percentage_error.py#L46): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals





Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

pranjaldatta commented:
Hey @SkafteNicki, I was working on #235. I just added the source code. But I started running into an issue with **pytest** and Ic couldn't figure out what the issue was. (Please ignore the docs/formatting test fails, I will fix them before the final PR ! )

So the issue is, 

### Issue

When I run `$ pytest tests/regression/test_mean_error.py -v
`

```shell
================================= test session starts =================================
platform linux -- Python 3.7.1, pytest-6.1.2, py-1.9.0, pluggy-0.13.1 -- /home/pranjal/miniconda3/envs/generalPT/bin/python
cachedir: .pytest_cache
rootdir: /home/pranjal/OSS/metrics, configfile: setup.cfg
collected 68 items                                                                    

tests/regression/test_mean_error.py::TestMeanError::test_mean_error_class[True-True-MeanSquaredError-mean_squared_error-mean_squared_error-preds0-target0-_single_target_sk_metric] PASSED [  1%]
tests/regression/test_mean_error.py::TestMeanError::test_mean_error_class[True-True-MeanSquaredError-mean_squared_error-mean_squared_error-preds1-target1-_multi_target_sk_metric] PASSED [  2%]
tests/regression/test_mean_error.py::TestMeanError::test_mean_error_class[True-True-MeanAbsoluteError-mean_absolute_error-mean_absolute_error-preds0-target0-_single_target_sk_metric] PASSED [  4%]
tests/regression/test_mean_error.py::TestMeanError::test_mean_error_class[True-True-MeanAbsoluteError-mean_absolute_error-mean_absolute_error-preds1-target1-_multi_target_sk_metric] PASSED [  5%]
tests/regression/test_mean_error.py::TestMeanError::test_mean_error_class[True-True-MeanSquaredLogError-mean_squared_log_error-mean_squared_log_error-preds0-target0-_single_target_sk_metric] PASSED [  7%]
tests/regression/test_mean_error.py::TestMeanError::test_mean_error_class[True-True-MeanSquaredLogError-mean_squared_log_error-mean_squared_log_error-preds1-target1-_multi_target_sk_metric] PASSED [  8%]
tests/regression/test_mean_error.py::TestMeanError::test_mean_error_class[True-True-MeanAbsolutePercentageError-mean_absolute_percentage_error-mean_absolute_percentage_error-preds0-target0-_single_target_sk_metric] 
```
It stalls at the **mean_absolute_percentage_error** test and I have to Keyboardnterrupt to exit the stall. I have no clue why this happening. I would really appreciate any help or pointers on how to solve this issue!

SkafteNicki commented on torchmetrics/functional/regression/mean_absolute_percentage_error.py:
```diff
@@ -0,0 +1,52 @@
+
+    _check_same_shape(preds, target)
+
+    #eps = torch.tensor(1.17e-07)# torch.tensor(np.finfo(np.float32).eps)
```
```suggestion

```

SkafteNicki commented on torchmetrics/functional/regression/mean_absolute_percentage_error.py:
```diff
@@ -0,0 +1,52 @@
+    return sum_abs_per_error / num_obs
+
+
+def mean_absolute_percentage_error(preds: torch.Tensor, target: torch.Tensor, eps: float= 1.17e-07) -> Tensor:
```
where did the `eps` value some from? seems very specific

pranjaldatta replied:
scikit-learn uses this value which happens to be a machine limit for np.float32 datatype. This value is calculated by ```np.finfo(np.float32).eps```. This particular value is the torch equivalent of it ```torch.finfo(torch.float32).eps```. I'll add this to the docs, to avoid confusion!

maximsch2 replied:
maybe specify it as a constant somewhere nearby with a comment?

pranjaldatta replied:
@maximsch2 I have done that in the latest commit and also added necessary references in the docs

SkafteNicki commented on torchmetrics/regression/mean_absolute_percentage_error.py:
```diff
@@ -0,0 +1,58 @@
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+    ):
```
you should be able to copy most of the docstring from the mean squared error function and just adjust with correct formula

pranjaldatta replied:
Done in latest commit

SkafteNicki commented on torchmetrics/regression/mean_absolute_percentage_error.py:
```diff
@@ -0,0 +1,58 @@
+        self.sum_abs_per_error += sum_abs_per_error
+        self.total += num_obs
+
+    def compute(self):
```
```suggestion
    def compute(self) -> Tensor:
```

SkafteNicki commented on torchmetrics/functional/regression/mean_absolute_percentage_error.py:
```diff
@@ -0,0 +1,52 @@
+
+
+def mean_absolute_percentage_error(preds: torch.Tensor, target: torch.Tensor, eps: float= 1.17e-07) -> Tensor:
+    """something"""
```
copy docstring from mean squared error function and adjust accordingly

pranjaldatta replied:
Done in latest commit

SkafteNicki commented on torchmetrics/regression/mean_absolute_percentage_error.py:
```diff
@@ -0,0 +1,58 @@
+        return _mean_absolute_percentage_error_compute(self.sum_abs_per_error, self.total)
+
+    @property
+    def is_differentiable(self):
```
```suggestion
    def is_differentiable(self) -> bool:
```

SkafteNicki reviewed (commented):
Hi @pranjaldatta, 
PR is looking really good. I am going to investigate the failing test one of the next days but here are some comments. In addition, please add the following:
* References in docs
* Entry in `changelog`

pranjaldatta commented:
> Hi @pranjaldatta,
> PR is looking really good. I am going to investigate the failing test one of the next days but here are some comments. In addition, please add the following:
> 
> * References in docs
> * Entry in `changelog`

Sure, I'll resolve them right away!

maximsch2 commented on torchmetrics/functional/regression/mean_absolute_percentage_error.py:
```diff
@@ -0,0 +1,51 @@
+# limitations under the License.
+from typing import Tuple
+
+import numpy as np
```
Let's not add the dependency on numpy

pranjaldatta replied:
Yup, I have applied this change with the latest commit!

maximsch2 commented on torchmetrics/functional/regression/mean_absolute_percentage_error.py:
```diff
@@ -13,21 +13,20 @@
 
-def _mean_absolute_percentage_error_update(preds: torch.Tensor, target: torch.Tensor, eps: torch.Tensor) -> Tuple[Tensor, int]:
+def _mean_absolute_percentage_error_update(
+    preds: torch.Tensor, target: torch.Tensor, epsilon: torch.Tensor
```
maybe consistently use `Tensor` vs `torch.Tensor` here? Shorter is probably better.

pranjaldatta replied:
I have fixed this in the latest commit! Thanks!

maximsch2 commented on torchmetrics/functional/regression/mean_absolute_percentage_error.py:
```diff
@@ -41,11 +40,34 @@ def _mean_absolute_percentage_error_compute(sum_abs_per_error: Tensor, num_obs:
 
-    eps = torch.tensor(eps)
-    sum_abs_per_error, num_obs = _mean_absolute_percentage_error_update(preds, target, eps)
+    epsilon = torch.tensor(epsilon)
```
does it need to be a tensor? Seems like maybe fine to have a regular float?

pranjaldatta replied:
So, I had initially used a normal float, but it would result in this error ```TypeError: max() received an invalid combination of arguments - got (float, Tensor)```. Hence the conversion to tensor!

maximsch2 replied:
OK, then you should at least do `tensor(epsilon, device=preds.device)` otherwise you'll get an error about tensors being on different devices once you start running this on GPU.

SkafteNicki replied:
after looking at the code again, it does not make sense to have `epsilon` as an argument. It is not something that the user should be able to control. It just to secure we do not divide by 0 :]

SkafteNicki commented:
Hi @pranjaldatta,
I have now fixed the test. The problem was with the `epsilon` parameter not being correctly transferred to the different devices. I have removed it as an user argument, since it is more of a numerical trick than a parameter that affects the metric.
Please take a look at the code and check that I have not completely butcher it :]

The PR looks done to me and is ready to be merged, but if you want to work more on it you can always think about adding the `multioutput` argument from the sklearn implementation.

pranjaldatta commented:
> Hi @pranjaldatta,
> I have now fixed the test. The problem was with the `epsilon` parameter not being correctly transferred to the different devices. I have removed it as an user argument, since it is more of a numerical trick than a parameter that affects the metric.
> Please take a look at the code and check that I have not completely butcher it :]
> 
> The PR looks done to me and is ready to be merged, but if you want to work more on it you can always think about adding the `multioutput` argument from the sklearn implementation.

Hey @SkafteNicki , thank you so much for your help! It works fine now!

Regarding `multioutput`, wouldn't it break consistency somewhat, since other metrics don't offer this feature as well?

SkafteNicki commented:
@pranjaldatta then lets not implement `multioutput` for now. We can always revisit if someone request the change :]
Going to mark this as ready for review.

Borda commented:
@pranjaldatta mind check the conflicts? :]

pranjaldatta commented:
@Borda resolved the merge conflicts!

Borda commented on torchmetrics/regression/mean_absolute_percentage_error.py:
```diff
@@ -0,0 +1,100 @@
+from torchmetrics.metric import Metric
+
+
+class MeanAbsolutePercentageError(Metric):
```
I feel the Percentage is confusing as the values shall be 0-100 but here are just relative
but Wiki has the same... https://en.wikipedia.org/wiki/Mean_absolute_percentage_error

Borda reviewed (commented):
seems we have the functional just badly named
https://github.com/PyTorchLightning/metrics/blob/233f0ebb606c5fe6749f3bada6a37f5e470c72cb/torchmetrics/functional/regression/mean_relative_error.py#L22
shall we reuse it or replace it with the new one?
@PyTorchLightning/core-metrics

Borda reviewed (changes requested):
still the duplicate has to be resolved https://github.com/PyTorchLightning/metrics/pull/248#pullrequestreview-664875965

pranjaldatta commented:
> still the duplicate has to be resolved [#248 (review)](https://github.com/PyTorchLightning/metrics/pull/248#pullrequestreview-664875965)

Hey @Borda, really sorry for the delay! Regarding the duplication, I really don't know which to keep. But like my two cents are, 

* The earlier function incorrectly named and hence I too missed it when I was looking for such a function.
* It doesn't have a class implementation.

Hence, this implementation seems more in-line with the rest of Mean-metrics. But again, that runs the risk of duplication. 

Please let me know if you require any modification in the current PR or the final decision regarding the choice!

SkafteNicki commented:
I am in favour of deprecating the old implementation and instead of the new one.

Borda commented:
@pranjaldatta also fine wit deprecating the old one

pranjaldatta commented:
> @pranjaldatta also fine wit deprecating the old one

Hey @Borda, just clarifying, you want me to push another commit deprecating the old one?

Borda commented:
> just clarifying, you want me to push another commit deprecating the old one?

yes, please :rabbit:

pranjaldatta commented:
Hey @Borda , @SkafteNicki , I was hoping to run something by you to ensure that it's the correct way of deprecation (I really don't want to mess up! :smile: ). So like to deprecate mean_relative_error.py, in **torchmetrics.utilities**, I add the following function,

```python
#torchmetrics/utilities/__init__.py

...

def _deprecation_warn_mean_relative_error() -> None:
    warn(
        "Function `mean_relative_error` was deprecated v0.3.3 and will be removed in v0.4." \
        "Use `mean_absolute_percentage_error` insteal.",
        DeprecationWarning
    )
```

and then in **torchmetrics/functional/regression/mean_relative_error.py**,
```python
#torchmetrics/functional/regression/mean_relative_error.py

def mean_relative_error(preds: Tensor, target: Tensor) -> Tensor:
    """
    Computes mean relative error

    Args:
        preds: estimated labels
        target: ground truth labels

    Return:
        Tensor with mean relative error

    Example:
        >>> from torchmetrics.functional import mean_relative_error
        >>> x = torch.tensor([0., 1, 2, 3])
        >>> y = torch.tensor([0., 1, 2, 2])
        >>> mean_relative_error(x, y)
        tensor(0.1250)

    """
    _deprecation_warn_mean_relative_error()
    sum_rltv_error, n_obs = _mean_relative_error_update(preds, target)
    return _mean_relative_error_compute(sum_rltv_error, n_obs
```
Is it the correct way of doing this? If not, I would be really grateful if you could direct me towards some previous resource/commit/PR as a reference!

pranjaldatta commented:
Hey @Borda, I have added the deprecation warning, I hope it's okay!

Borda commented on torchmetrics/functional/regression/mean_absolute_percentage_error.py:
```diff
@@ -0,0 +1,68 @@
+
+def _mean_absolute_percentage_error_update(preds: Tensor,
+                                           target: Tensor,
+                                           epsilon: float = 1.17e-06) -> Tuple[Tensor, int]:
```
```suggestion
                                           epsilon: float = 1.17e-06,) -> Tuple[Tensor, int]:
```

Borda commented on torchmetrics/functional/regression/mean_relative_error.py:
```diff
@@ -51,5 +52,9 @@ def mean_relative_error(preds: Tensor, target: Tensor) -> Tensor:
 
     """
+    warn(
+        "Function `mean_relative_error` was deprecated v0.3.3 and will be removed in v0.4."
```
```suggestion
        "Function `mean_relative_error` was deprecated v0.4 and will be removed in v0.5."
```

Borda commented on torchmetrics/functional/regression/mean_relative_error.py:
```diff
@@ -51,5 +52,9 @@ def mean_relative_error(preds: Tensor, target: Tensor) -> Tensor:
+        "Function `mean_relative_error` was deprecated v0.3.3 and will be removed in v0.4."
+        "Use `mean_absolute_percentage_error` instead.", DeprecationWarning
+    )
     sum_rltv_error, n_obs = _mean_relative_error_update(preds, target)
     return _mean_relative_error_compute(sum_rltv_error, n_obs)
```
use the `_mean_absolute_percentage_error_update` and `_mean_absolute_percentage_error_compute` instead

pranjaldatta replied:
Applied the changes via the latest commit!

Borda commented on tests/regression/test_mean_error.py:
```diff
@@ -47,14 +58,14 @@
     sk_preds = preds.view(-1).numpy()
     sk_target = target.view(-1).numpy()
-    res = sk_fn(sk_preds, sk_target)
+    res = sk_fn(sk_target, sk_preds)
```
why this switch?

pranjaldatta replied:
So some time back I had [this issue](https://github.com/PyTorchLightning/metrics/pull/248#issuecomment-841232277) and to resolve this, @SkafteNicki came up with [this switch](https://github.com/PyTorchLightning/metrics/pull/248#issuecomment-842976718), and since then all tests are passing.

Borda replied:
ok, can you and a comment/explanation because otherwise will have the same question whenever anyone will be touching

pranjaldatta replied:
Done!

Borda commented on torchmetrics/functional/regression/mean_relative_error.py:
```diff
@@ -50,6 +40,14 @@ def mean_relative_error(preds: Tensor, target: Tensor) -> Tensor:
+    warn(
+        "Function `mean_relative_error` was deprecated v0.4 and will be removed in v0.5."
+        "Use `mean_absolute_percentage_error` instead.",
+        DeprecationWarning,
```
```suggestion
        DeprecationWarning
```

Borda commented:
@pranjaldatta thank you for your patience! :rabbit:

pranjaldatta commented:
> @pranjaldatta thank you for your patience!

Always a pleasure to contribute to this community in whatever small way possible!
